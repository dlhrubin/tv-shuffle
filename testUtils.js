import { fireEvent } from '@testing-library/react';
import axiosMock from 'axios';
import successSearch from './__fixtures__/successSearch';
import successShowInfo from './__fixtures__/successShowInfo';
import noPosterSearch from './__fixtures__/noPosterSearch';
import noPosterShowInfo from './__fixtures__/noPosterShowInfo';
import notFoundSearch from './__fixtures__/notFoundSearch';
import noDataSearch from './__fixtures__/noDataSearch';
import noDataShowInfo from './__fixtures__/noDataShowInfo';

jest.mock('axios');

// Mock successful API call (returns show)
export const successfulCall = (getByTestId) => {
  fireEvent.change(getByTestId('search-bar'), { target: { value: 'how i met' } });
  axiosMock.get.mockResolvedValueOnce(successSearch);
  axiosMock.get.mockResolvedValueOnce(successShowInfo);
  fireEvent.submit(getByTestId('search-form'));
};

// Mock successful API call (returns show but no poster)
export const noPosterCall = (getByTestId) => {
  fireEvent.change(getByTestId('search-bar'), { target: { value: 'days of de' } });
  axiosMock.get.mockResolvedValueOnce(noPosterSearch);
  axiosMock.get.mockResolvedValueOnce(noPosterShowInfo);
  fireEvent.submit(getByTestId('search-form'));
};

// Mock unsuccessful API call (show title not found)
export const notFoundCall = (getByTestId) => {
  fireEvent.change(getByTestId('search-bar'), { target: { value: 'abcd' } });
  axiosMock.get.mockResolvedValueOnce(notFoundSearch);
  fireEvent.submit(getByTestId('search-form'));
};

// Mock unsuccessful API call (insufficient data found for show)
export const noDataCall = (getByTestId) => {
  fireEvent.change(getByTestId('search-bar'), { target: { value: 'days of' } });
  axiosMock.get.mockResolvedValueOnce(noDataSearch);
  axiosMock.get.mockResolvedValueOnce(noDataShowInfo);
  fireEvent.submit(getByTestId('search-form'));
};
