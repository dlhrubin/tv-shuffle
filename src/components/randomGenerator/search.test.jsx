import React from 'react';
import { render, cleanup, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { successfulCall, notFoundCall, noDataCall } from '../../../testUtils';
import Search from './search';

afterEach(cleanup);

describe('Search', () => {
  it('renders without crashing', () => {
    const { asFragment } = render(<Search />);
    expect(asFragment(<Search />)).toMatchSnapshot();
  });
  describe('TV show search bar', () => {
    it('queries API with user-entered search query', async () => {
      const { getByTestId } = render(<Search />);
      successfulCall(getByTestId);
      await waitFor(() => {
        expect(getByTestId('search-bar').value).toEqual('How I Met Your Mother');
      });
    });
    it('returns an error if show title not found', async () => {
      const { getByTestId } = render(<Search />);
      expect(getByTestId('error-message')).toHaveTextContent('');
      notFoundCall(getByTestId);
      await waitFor(() => {
        expect(getByTestId('search-bar').value).toEqual('abcd');
        expect(getByTestId('error-message')).toHaveTextContent('Show title not found');
      });
    });
    it('returns an error if no data available for a show', async () => {
      const { getByTestId } = render(<Search />);
      noDataCall(getByTestId);
      await waitFor(() => {
        expect(getByTestId('search-bar').value).toEqual('Days of Our Lives');
        expect(getByTestId('error-message')).toHaveTextContent('No data available for this title');
      });
    });
  });
  describe('Season dropdown menu', () => {
    it('is enabled and populates with options when search bar returns result', async () => {
      const { container, getByTestId, getByText } = render(<Search />);
      successfulCall(getByTestId);
      await waitFor(() => {
        expect(getByTestId('dropdown')).toBeEnabled();
        // Check for option elements
        expect(container).toContainElement(getByText('All'));
        for (let i = 1; i <= 9; i++) {
          expect(container).toContainElement(getByText(i.toString()));
        }
      });
    });
  });
});
