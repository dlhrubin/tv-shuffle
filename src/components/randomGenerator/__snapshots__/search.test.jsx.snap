// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders without crashing 1`] = `
<DocumentFragment>
  <section
    class="search"
    id="search"
  >
    <form
      data-testid="search-form"
    >
      <label
        for="find-show"
      >
        <span>
          Show
        </span>
        <div>
          <div
            class="input-container"
          >
            <input
              autocomplete="off"
              data-testid="search-bar"
              id="find-show"
              type="search"
              value=""
            />
            <button
              aria-label="Search"
              type="submit"
            >
              <i
                class="fas fa-search"
              />
            </button>
          </div>
          <span
            class="error"
            data-testid="error-message"
          />
        </div>
      </label>
    </form>
    <form>
      <label
        for="select-season"
      >
        <span>
          Season (optional)
        </span>
        <select
          data-testid="dropdown"
          disabled=""
          id="select-season"
        >
          00
        </select>
      </label>
    </form>
  </section>
</DocumentFragment>
`;
